src/Bicep.Cli/Services/TemplateEvaluatorService.cs

```diff
@@ -0,0 +1,208 @@
+
+        private static string GetResourceId(string scopeString, TemplateResource resource)
+        {
+            var typeSegments = resource.Type.Value.Split('/');
+            var nameSegments = resource.Name.Value.Split('/');
+
+            var types = new[] { typeSegments.First() }
+                .Concat(typeSegments.Skip(1).Zip(nameSegments, (type, name) => $"{type}/{name}"));
+
+            return $"{scopeString}providers/{string.Join('/', types)}";
+        }
+
+        private static void ProcessTemplateLanguageExpressions(Template template, EvaluationConfiguration config, TemplateDeploymentScope deploymentScope)
+        {
+            var scopeString = deploymentScope switch
+            {
+                TemplateDeploymentScope.Tenant => "/",
+                TemplateDeploymentScope.ManagementGroup => $"/providers/Microsoft.Management/managementGroups/{config.ManagementGroup}/",
+                TemplateDeploymentScope.Subscription => $"/subscriptions/{config.SubscriptionId}/",
+                TemplateDeploymentScope.ResourceGroup => $"/subscriptions/{config.SubscriptionId}/resourceGroups/{config.ResourceGroup}/",
+                _ => throw new InvalidOperationException(),
+            };
+
+            var resourceLookup = template.Resources.ToOrdinalInsensitiveDictionary(x => GetResourceId(scopeString, x));
+
+            var evaluationContext = TemplateEngine.GetExpressionEvaluationContext(config.ManagementGroup, config.SubscriptionId, config.ResourceGroup, template, null);
+            var defaultEvaluateFunction = evaluationContext.EvaluateFunction;
+            evaluationContext.EvaluateFunction = (FunctionExpression functionExpression, FunctionArgument[] parameters, TemplateErrorAdditionalInfo additionalInfo) =>
+            {
+                if (functionExpression.Function.StartsWithOrdinalInsensitively(LanguageConstants.ListFunctionPrefix) && config.OnListFunc is not null)
+                {
+                    return config.OnListFunc(
+                        functionExpression.Function,
+                        parameters[0].Token.ToString(),
+                        parameters[1].Token.ToString(),
+                        parameters.Length > 2 ? parameters[2].Token : null);
+                }
+
+                if (functionExpression.Function.EqualsOrdinalInsensitively("reference"))
+                {
+                    var resourceId = parameters[0].Token.ToString();
+                    var apiVersion = parameters.Length > 1 ? parameters[1].Token.ToString() : null;
+                    var fullBody = parameters.Length > 2 ? parameters[2].Token.ToString().EqualsOrdinalInsensitively("Full") : false;
+
+                    if (apiVersion is not null && config.OnReferenceFunc is not null)
+                    {
+                        return config.OnReferenceFunc(resourceId, apiVersion, fullBody);
+                    }
+
+                    if (resourceLookup.TryGetValue(resourceId, out var foundResource) &&
```

Comment:
We also need to handle the case where the symbolic name codegen is enabled, and the parameter of `reference` function call is a resource symbolic name instead of a resource ID.

Reply:
Discussed offline. If we set the `onGetTemplateReference` parameter for `TemplateEngine.GetExpressionEvaluationContext`, symbolic name reference will be handled automatically. We just need to move the resource reference lookup logic to `onGetTemplateReference` to handle references with resource IDs.

Reply:
The change can be made in a follow-up PR.